src/server/controllers/s3.js

```diff
@@ -57,4 +41,133 @@ module.exports = function s3 (config) {
+
+      if (typeof key !== 'string') {
+        return res.status(400).json({ error: 's3: the object key must be passed as a query parameter. For example: "?key=abc.jpg"' })
+      }
+
+      client.listParts({
+        Bucket: config.bucket,
+        Key: key,
+        UploadId: uploadId
+      }, (err, data) => {
+        if (err) {
+          next(err)
+          return
+        }
+
+        // TODO Account for `data.IsTruncated` when there are more than 1000 parts
+        res.json(data.Parts)
+      })
+    })
+    .get('/multipart/:uploadId/:partNumber', (req, res, next) => {
+      const client = req.uppy.s3Client
+      const { uploadId, partNumber } = req.params
+      const { key } = req.query
+
+      if (typeof key !== 'string') {
+        return res.status(400).json({ error: 's3: the object key must be passed as a query parameter. For example: "?key=abc.jpg"' })
+      }
+      if (!parseInt(partNumber, 10)) {
+        return res.status(400).json({ error: 's3: the part number must be a number between 1 and 10000.' })
+      }
+
+      client.getSignedUrl('uploadPart', {
+        Bucket: config.bucket,
+        Key: key,
+        UploadId: uploadId,
+        PartNumber: partNumber,
+        Body: '',
+        Expires: ms('5 minutes') / 1000
+      }, (err, url) => {
+        if (err) {
+          next(err)
+          return
+        }
+        res.json({ url })
+      })
+    })
+    .delete('/multipart/:uploadId', (req, res, next) => {
+      const client = req.uppy.s3Client
+      const { uploadId } = req.params
+      const { key } = req.query
+
+      if (typeof key !== 'string') {
+        return res.status(400).json({ error: 's3: the object key must be passed as a query parameter. For example: "?key=abc.jpg"' })
+      }
+
+      client.abortMultipartUpload({
+        Bucket: config.bucket,
+        Key: key,
+        UploadId: uploadId
+      }, (err, data) => {
+        if (err) {
+          next(err)
+          return
+        }
+        res.json({})
+      })
+    })
+    .post('/multipart/:uploadId/complete', (req, res, next) => {
+      const client = req.uppy.s3Client
+      const { uploadId } = req.params
+      const { key } = req.query
+      const { parts } = req.body
+
+      if (typeof key !== 'string') {
+        return res.status(400).json({ error: 's3: the object key must be passed as a query parameter. For example: "?key=abc.jpg"' })
+      }
+      if (!Array.isArray(parts) || !parts.every(isValidPart)) {
+        return res.status(400).json({ error: 's3: `parts` must be an array of {ETag, PartNumber} objects.' })
+      }
+
+      client.completeMultipartUpload({
+        Bucket: config.bucket,
+        Key: key,
+        UploadId: uploadId,
+        MultipartUpload: {
+          Parts: parts
+        }
+      }, (err, data) => {
+        if (err) {
+          next(err)
+          return
+        }
+        res.json({
+          location: data.Location
+        })
+      })
+    })
+}
```

Comment:
having all these endpoints makes me now wonder if we should merge s3 into a provider like the others somehow

Reply:
It is a bit unorganised like this, yes… Maybe we can define a class earlier in the file, with instance methods for the routes? I'm not sure what merging it into a provider would look like since the S3 routes do very different things compared to the provider routes.

Reply:
makes sense! I moved them into functions and added some comments about what they do and what inputs/outputs they expect